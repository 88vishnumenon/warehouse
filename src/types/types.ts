export interface Product{
    id:string,
    name:string,
    articles:Array<ProductArticle>
}

export interface ProductArticle{
    id:string,
    amountRequired:number
}

export interface Article {
  id: string;
  name: string;
  amountInStock: number;
}

export interface Sale {
  id: string;
  createdAt: string;
  productId: string;
  amountSold:number
}

export interface TableData {
  value: string | number;
}

export interface ProductSold {
  id: string;
  createdAt: string;
  productId: string;
  amountSold: number;
}