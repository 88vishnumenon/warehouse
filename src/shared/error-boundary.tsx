

import { makeStyles } from "@material-ui/styles";
import { Theme } from "@mui/material";


const useStyles = makeStyles((theme: Theme) => ({
  loadingWrapper: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    height: "40%",
  },

  error: {
    backgroundImage: `url(/assets/images/error.jpeg)`,
    backgroundPosition: "center",
    backgroundSize: "cover",
    backgroundRepeat: "no-repeat",
    margin: "0 auto",
    height: "40vh",
    width: "30%",
    display: "flex",
  },
}));


const ErrorComponent: React.FC<{}> = (props) => {

  const classes = useStyles();
  return <div className={classes.error}></div>;
}

export default ErrorComponent;
