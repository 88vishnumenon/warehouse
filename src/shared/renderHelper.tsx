import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@material-ui/core/styles";
import { Provider } from "react-redux";
import {reducer} from "../store/reducer";
import { createStore } from "redux";
import customTheme from "../assets/theme/custom-theme";


// can improve
// 1.change redux state from any to type
export const renderWithProviders = (
    component: JSX.Element,
    reduxState: any
) => {
    const UI = (
        <ThemeProvider theme={customTheme}>
            {component}
        </ThemeProvider>
    );
    if (reduxState) {
        const store = createStore(reducer,reduxState);
        const Wrapper:any = ({ children }: { children: React.ReactChildren }) => {
            return <Provider store={store}>{children}</Provider>;
        };
        return render(UI, { wrapper: Wrapper });
    }
    return render(UI);
};



