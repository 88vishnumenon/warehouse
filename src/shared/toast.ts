import iZtoast from "izitoast";
import "izitoast/dist/css/iziToast.min.css";

const toast = {
  error: (message: string, title = "Error") => {
    return iZtoast.error({
      title: title,
      message: message,
      position: "topRight",
    });
  },
  success: (message: string, title = "Success") => {
    return iZtoast.success({
      title: title,
      message: message,
      position: "topRight",
    });
  },
};

export default toast;
