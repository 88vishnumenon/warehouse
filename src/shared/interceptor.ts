import axios from "axios";
import toast from "./toast";

import { AxiosError, AxiosRequestConfig, AxiosResponse } from "axios";

const customAxios = axios.create({
  baseURL: `http://localhost:7000`,
});

const onRequest = (config: AxiosRequestConfig): AxiosRequestConfig => {
  return config;
};

const onRequestError = (error: AxiosError): Promise<AxiosError> => {
  console.error(`[request error] [${JSON.stringify(error)}]`);
  return Promise.reject(error);
};

const onResponse = (response: AxiosResponse): AxiosResponse => {
  return response;
};

const onResponseError = (error: AxiosError): Promise<AxiosError> => {
    // not showing the error when the error is of cancelled request
    if (error.message != "axios request cancelled"){
      toast.error("Oops, something went wrong!");
    } 
    return Promise.reject(error);
};

customAxios.interceptors.request.use(
  (request) => onRequest(request),
  (error) => onRequestError(error)
);

customAxios.interceptors.response.use(
  (response) => onResponse(response),
  (error) => onResponseError(error)
);

export default customAxios;
