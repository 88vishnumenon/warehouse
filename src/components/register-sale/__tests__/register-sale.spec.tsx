


import { fireEvent, within } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { renderWithProviders } from "../../../shared/renderHelper";
import { Article, Product } from "../../../types/types";
import RegisterSale from "../register-sale";

// mocks
export const MOCK_STATE = {
    products: [
        {
            "id": "a269a247-0d38-4b47-9630-79c9ae545b68",
            "name": "Dining Chair111111",
            "articles": [
                {
                    "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
                    "amountRequired": 4
                },
                {
                    "id": "831b92b8-677b-42cc-a585-335ea4ccccb6",
                    "amountRequired": 1
                },
                {
                    "id": "addc65a8-c759-41d8-a18a-89fe446ad484",
                    "amountRequired": 8
                }
            ]
        },

    ],
    articles: [
        {
            "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
            "name": "Leg",
            "amountInStock": 400
        },
        {
            "id": "addc65a8-c759-41d8-a18a-89fe446ad484",
            "name": "Screw",
            "amountInStock": 600
        },
        {
            "id": "831b92b8-677b-42cc-a585-335ea4ccccb6",
            "name": "Seat",
            "amountInStock": 200
        },
        {
            "id": "6892b98b-9b87-4520-9a9e-7528f1d78cb4",
            "name": "Table Top",
            "amountInStock": 100
        }
    ]

};

//mock-apis
// const sellProductMock= jest.fn(() =>{})
// jest.mock('../../../services/services', () => ({
//     sellProduct: sellProductMock
// }));

const mockSellProduct = jest.fn();

jest.mock('../../../services/services', () => ({
    sellProduct: () => mockSellProduct, 
}));




const renderRegisterSaleList = (
    state: {
        products: Array<Product>,
        articles: Array<Article>
    }
) => {
    return renderWithProviders(
        <RegisterSale></RegisterSale>,
        state
    );
};

describe("<RegisterSale /> test utils", () => {
    test("should match snapshot", async () => {
        const { asFragment } = renderRegisterSaleList(MOCK_STATE);
        expect(asFragment()).toMatchSnapshot();
    })

    test("should submit button is disabled",async()=>{
        const { findByTestId } = renderRegisterSaleList(MOCK_STATE);
        const submitBtn = await findByTestId("submit-btn");
        expect(submitBtn).toBeDisabled();
    })

   test("should all elements be in the document",async()=>{
       const { findByTestId } = renderRegisterSaleList(MOCK_STATE);
       const saleType = (await findByTestId("sale-types")) as HTMLSelectElement;
       const saleItem = (await findByTestId("sale-item-types")) as HTMLSelectElement;
       const amount = await (await findByTestId("amount")).querySelector('.MuiInputBase-input') as HTMLInputElement;

       expect(saleType).toBeInTheDocument();
       expect(saleItem).toBeInTheDocument();
       expect(amount).toBeInTheDocument();


   })

    test("should submit btn be enabled after all values are entered", async () => {
        const { findByTestId,getByTestId } = renderRegisterSaleList(MOCK_STATE);
        const saleType = await (await findByTestId("sale-types"))
        const saleItem = (await findByTestId("sale-item-types")) ;
        const amount = await (await findByTestId("amount")).querySelector('.MuiInputBase-input') as HTMLInputElement;
        const submitBtn = await findByTestId("submit-btn");


        // select value in autocomplete products
        const typeInput = await getByTestId("sale-types").querySelector('.MuiInputBase-input') as HTMLInputElement;
        saleType.focus();
    
        fireEvent.change(typeInput, { target: { value: 'p' } })
        fireEvent.keyDown(saleType, { key: 'ArrowDown' })
        fireEvent.keyDown(saleType, { key: 'Enter' })

        //select value in autocomplete Items
        const typeItem = await getByTestId("sale-item-types").querySelector('.MuiInputBase-input') as HTMLInputElement;
        saleItem.focus();
     
        fireEvent.change(typeItem, { target: { value: 'd' } })
        fireEvent.keyDown(saleItem, { key: 'ArrowDown' })
        fireEvent.keyDown(saleItem, { key: 'Enter' })


        fireEvent.change(amount, { target: { value: 10 } });

        expect(submitBtn).not.toBeDisabled();


    })


    // test("should api call be made on submit button", async () => {
    //     const { findByTestId, getByTestId } = renderRegisterSaleList(MOCK_STATE);
    //     const saleType = await (await findByTestId("sale-types"))
    //     const saleItem = (await findByTestId("sale-item-types"));
    //     const amount = await (await findByTestId("amount")).querySelector('.MuiInputBase-input') as HTMLInputElement;
    //     const submitBtn = await findByTestId("submit-btn");


    //     // select value in autocomplete products
    //     const typeInput = await getByTestId("sale-types").querySelector('.MuiInputBase-input') as HTMLInputElement;
    //     saleType.focus();

    //     fireEvent.change(typeInput, { target: { value: 'p' } })
    //     fireEvent.keyDown(saleType, { key: 'ArrowDown' })
    //     fireEvent.keyDown(saleType, { key: 'Enter' })

    //     //select value in autocomplete Items
    //     const typeItem = await getByTestId("sale-item-types").querySelector('.MuiInputBase-input') as HTMLInputElement;
    //     saleItem.focus();

    //     fireEvent.change(typeItem, { target: { value: 'd' } })
    //     fireEvent.keyDown(saleItem, { key: 'ArrowDown' })
    //     fireEvent.keyDown(saleItem, { key: 'Enter' })


    //     fireEvent.change(amount, { target: { value: 10 } });

    //     fireEvent.click(submitBtn);
    //     const spy = jest.spyOn("../../../services/services", "sellProduct");

    //     expect(spy).toHaveBeenCalled();



    // })
})