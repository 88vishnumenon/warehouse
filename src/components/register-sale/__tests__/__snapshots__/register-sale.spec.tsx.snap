// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterSale /> test utils should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="makeStyles-registerSaleWrapper-1"
  >
    <form
      class="makeStyles-registerSaleForm-2"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon makeStyles-saleInputs-3 css-viou3o-MuiAutocomplete-root"
        data-testid="sale-types"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="sale-types"
            id="sale-types-label"
          >
            Type
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="sale-types"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1g8cwq9"
              >
                <span>
                  Type
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon makeStyles-saleInputs-3 css-viou3o-MuiAutocomplete-root"
        data-testid="sale-item-types"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="sale-item-types"
            id="sale-item-types-label"
          >
            Items
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="sale-item-types"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1g8cwq9"
              >
                <span>
                  Items
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-saleInputs-3 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-testid="amount"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="amount"
          id="amount-label"
        >
          Amount
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="amount"
            min="1"
            type="number"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1g8cwq9"
            >
              <span>
                Amount
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <section
        class="makeStyles-btnLoadingWrapper-8"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled makeStyles-submitBtn-5 css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="submit-btn"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Submit
        </button>
      </section>
    </form>
  </section>
</DocumentFragment>
`;
